h2/src/main/org/h2/mvstore/Page.java

```diff
@@ -967,166 +869,517 @@ public void removePage() {
-                    continue;
+        private Object values[];
+
+        Leaf(MVMap<?, ?> map) {
+            super(map);
+        }
+
+        private Leaf(MVMap<?, ?> map, Leaf source) {
+            super(map, source);
+            this.values = source.values;
+        }
+
+        Leaf(MVMap<?, ?> map, Object keys[], Object values[]) {
+            super(map, keys);
+            this.values = values;
+        }
+
+        @Override
+        public int getNodeType() {
+            return PAGE_TYPE_LEAF;
+        }
+
+        @Override
+        public Page copy(MVMap<?, ?> map) {
+            return new Leaf(map, this);
+        }
+
+        @Override
+        public Page getChildPage(int index) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public Page getChildPageIfLoaded(int index) { throw new UnsupportedOperationException(); }
+
+        @Override
+        public long getChildPagePos(int index) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public Object getValue(int index) {
+            return values[index];
+        }
+
+        @Override
+        @SuppressWarnings("SuspiciousSystemArraycopy")
+        public Page split(int at) {
+            assert !isSaved();
+            int b = getKeyCount() - at;
+            Object bKeys[] = splitKeys(at, b);
+            Object bValues[] = createValueStorage(b);
+            if(values != null) {
+                Object aValues[] = createValueStorage(at);
+                System.arraycopy(values, 0, aValues, 0, at);
+                System.arraycopy(values, at, bValues, 0, b);
+                values = aValues;
+            }
+            Page newPage = create(map, bKeys, bValues, null, b, 0);
+            if(isPersistent()) {
+                recalculateMemory();
+            }
+            return newPage;
+        }
+
+        @Override
+        public long getTotalCount() {
+            return getKeyCount();
+        }
+
+        @Override
+        long getCounts(int index) {
+            throw new UnsupportedOperationException();
+        }
+        @Override
+        public void setChild(int index, Page c) {
+            throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public Object setValue(int index, Object value) {
+            DataType valueType = map.getValueType();
+            values = values.clone();
+            Object old = setValueInternal(index, value);
+            if(isPersistent()) {
+                addMemory(valueType.getMemory(value) -
+                            valueType.getMemory(old));
+            }
+            return old;
+        }
+
+        private Object setValueInternal(int index, Object value) {
+            Object old = values[index];
+            values[index] = value;
+            return old;
+        }
+
+        @Override
+        @SuppressWarnings("SuspiciousSystemArraycopy")
```

Comment:
`System.arraycopy()` is not used here too.

Reply:
Not yet, back-port hiccup